### PR TITLE
Add landau energy conservation test

### PR DESCRIPTION
**Pull request by ckendrick**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

This adds a test that runs the Landau PDE for 100 time steps using IMEX and tests the energy conservation. This test should help detect issues with any changes to IMEX in the future.

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [ ] Bugfix
- [ ] New feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

## What systems has this change been tested on?

Ubuntu 20.04

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

### CODE REVIEW

File: src/time_advance_tests.cpp [hunk 1]
```diff
@@ -1237,3 +1237,135 @@ TEMPLATE_TEST_CASE("implicit time advance - continuity 2", "[time_advance]",
...
+    if (i == 0)
+    {
+      E_pot_initial = E_pot;
+      E_kin_initial = E_kin;
+    }
+
+    // calculate the absolute relative total energy
+    TestType E_relative =
+        std::fabs((E_pot + E_kin) - (E_pot_initial + E_kin_initial));
+    REQUIRE(E_relative <= 1.0e-9);
```
Comment (mkstoyanov) [inline src/time_advance_tests.cpp]: This will fail in single precision

File: src/time_advance_tests.cpp [hunk 2]
```diff
@@ -1237,3 +1237,135 @@ TEMPLATE_TEST_CASE("implicit time advance - continuity 2", "[time_advance]",
...
+
+      TestType integral = grid_spacing * sum;
+
+      return integral;
+    };
+
+    // compute the E potential and kinetic energy
+    fk::vector<TestType> E_field_sq(pde->E_field);
+    std::transform(E_field_sq.begin(), E_field_sq.end(), E_field_sq.begin(),
+                   E_field_sq.begin(), std::multiplies<TestType>());
```
Comment (quantumsteve) [inline src/time_advance_tests.cpp]:
```suggestion
                   E_field_sq.begin(), std::multiplies<>());
```

Comment (quantumsteve) [inline src/time_advance_tests.cpp]: A range-based for loop that squares each term might be clearer.

Comment (mkstoyanov) [reply src/time_advance_tests.cpp]: @ckendrick this is actually horrible code. In a general context the transforms templates can be considered `good` but in a math context, this is really really bad.

Math formulas and simple indexing are easy to read, we are doing math and we should write code close to math formulas:
```
for(int i=0; i<E_field_sq.size(); i++)
  E_field_sq[i] *= E_field_sq[i];
```
or as @quantumsteve suggested
```
for(auto &e : E_field_sq)
  e = e * e;
```

**Review by mkstoyanov — APPROVED** [review]

**Pull request merged**